### DISCUSSION TITLE
2nd KRATOS COMMUNITY MEETING

### DISCUSSION THREAD

**Discussion by maceligueta** [KratosMultiphysics/Kratos#8430] [General]
# **2nd KRATOS COMMUNITY MEETING**

### https://unipd.zoom.us/j/81286277636

This meeting follows the series started with the [1st Kratos Community Meeting](https://github.com/KratosMultiphysics/Kratos/discussions/8239)

## ADDRESSED TOPICS:
First, @philbucher will address the following technical topics:
- How the CI (continuous integration) works => the checks that have to be passed before a PR can be merged. In the same note also go into how the testing is set up and working
- The future of shared memory parallelization in Kratos: The transition from OpenMP to C++ Threads (see [here](https://github.com/KratosMultiphysics/Kratos/wiki/Parallel-Utilities))
- The future of C++ testing in Kratos

Second, a question proposed by the community will be answered by a selected speaker. This question will be the most upvoted among those posted by the pasticipants in this discussion.

### •	Post your questions, doubts, issues and suggestions as contributions to this discussion before 12th March 2021

•	Use **‘thumb up’ or upvoting** to support previously written questions. Some of the questions are ported from the previous meeting, because there was not enough time to address all of them. The most popular questions will have priority

## AGENDA for 25th March 2021 at 9.00 (CET)

▪	Welcome, introduction and presentation of the selected questions (5 mins)

▪	CI, future of shared memory parallelization and C++ testing  (10 mins) 

▪	Second question is addressed  (10 mins) 

▪	Open discussion by ‘raising hands’ system (5 mins)

## VIDEO AND MATERIAL OF THE MEETING

Video of the meeting [Link to video](https://mediaspace.unipd.it/media/2nd+KRATOS+Community+Meeting/1_27ebjrnb)
Pdf of the presentation by @philbucher [Link to slides](https://drive.google.com/file/d/1Ktbcja214ZXrHZqjqZuTAm1yLr7DpJew/view?usp=sharing)

@maceligueta @antonialarese @rolandwuechner

**Comment by maceligueta** [comment]
Question proposed by @riccardotosi :
What is your opinion of Kratos with respect to other open source software (if you have any)? Why should one prefer Kratos over other open-source software?
From this question in Linkedin you have a list of some open-source software:

Basilisk (http://basilisk.fr/)
code_saturne (https://lnkd.in/dte92kG)
Elmer (https://lnkd.in/deX9SJS)
FEniCS (https://fenicsproject.org)
Fluidity (https://lnkd.in/dm68gZb)
Kratos (https://lnkd.in/d-a2ENK)
Nektar++ (https://www.nektar.info)
OpenFOAM (https://www.openfoam.com)
Sparselizard (https://lnkd.in/dPjnJH9)
SU2 (https://su2code.github.io)

**Reply by antonialarese** [reply]
@KratosMultiphysics/all Does anyone have experience with one of the open-source codes mentioned by @riccardotosi? If so, please let us know it by the end of the week (better by email at antonia.larese@unipd.it).
This way we can organize the contributions to the 2nd  Kratos Community meeting.
Thanks!

**Comment by maceligueta** [comment]
Question proposed by @AlejandroCornejo : 
Another useful point could be about the encapsulation of Kratos in an executable file to be delivered easily to early users/students withour having to compile Kratos inside each PC which can be costly and time consuming

**Reply by philbucher** [reply]
I would go with the wheels for this rather than to come up with our own solution
this is the default solution for installing python packages and super easy => `pip install KratosMultiphysics`

**Reply by jrubiogonzalez** [reply]
Encapsulating in an executable binary is quite straightforward if we have a unified entry point, we are doing so. I think for a regular user that is not a programmer and just an end user would be the best option (and does not collide with the wheels option)

**Comment by ddiezrod** [comment]
I have one doubt. Over the years, I have heard several times that Kratos can not directly compete in terms of speed with "x" commercial software as we are more focused on being very flexible. I understand this is due to some "core" decisions, so I'd like to know more exactly which those sacrifices are that were made in favor of flexibility and if they could be reverted/improved in the future.

**Reply by philbucher** [reply]
IMO if we would be competitive with commercial solvers then they wouldn't do their job correct :D

More serious:
Developing for several years already I think that optimizing for speed is one of the most complicated tasks. It requires lots of deep understanding of programming itself as well as an understanding of the physics behind.
I am guessing that commercial solver spend a lot of effort on making their things fast, and that is time and expertise that only few of us have.
Hence I don't think that we necessarily made sacrifices but don't have the manpower to improve (even more)

**Reply by jrubiogonzalez** [reply]
Hi, IMO I don't know if this is a conscious option, anyway knowing what are the limitations/bottlenecks is the first step towards improving. 

We will probably not achieve top class speed, but maybe being somewhere reasonably close. Right now our benchmarks are that out solutions are somwhere x2 - x4 slower.

**Reply by ddiezrod** [reply]
Just to be clear, I'm not saying that we should be as fast as a commercial code. (Those guys have to earn a living : P) . I might be wrong, but my impression in some discussions was that we have some limitations in Kratos that are "well known" and somehow conscious decisions. But if I think in, for example, a fluid problem where pretty much all the cost go directly to build and solve I think there is no a big reason why we should be much slower.

Anyway, probably what you say is closer to the truth and so we should try to find bottlenecks elsewhere. I'd just like to have the whole picture.

**Reply by loumalouomega** [reply]
We are faster than many research codes that target wide types of problems (I know a guy who implemented a super fast NS solver, that directly passed the data to PETSc, but only worked for tetra, and only for that kind of problems)

**Reply by pooyan-dadvand** [reply]
Seems that I agree with both sides of the discussion! So at least they are not orthogonal in my space! :stuck_out_tongue: 

On one side the objective of the Kratos was multiphysics and flexibility and that was the excuse for us for years to go more features and keep the performance reasonably good. (we used to be less than x2 slower than single purposed codes) Our main luck is the fact that in our simulations the linear solver is the most time-consuming part and we use the best of the market. So the inefficiency in other parts usually is not very crucial. Meanwhile, the flexibility of the Kratos make it the fastest in many solutions in which there are many disciplines involved. 
 
On the other side, I see the concern of the @ddiezrod about being bound to go considerably slower than the single purpose codes. (Regardless that they are commercial or not) Please note that now we are 4x slower than other software. Also, we have solutions that linear solver is not involved (like DEM and explicit) where the flexible structure of Kratos imposes more overhead. 

So maybe, it is time to see if we can improve the situation. At least, as a design exercise before any real change or refactor. For example how we can bound the flexibility in some cases to improve a global performance? An example of this exercise in the past resulted in an important change of the SolutionStepValues which should be configured from the beginning. It used to be completely dynamic, but now we know perfectly that maintaining that in multithreading would be impossible.

**Reply by jrubiogonzalez** [reply]
They are not orthogonal, Kratos main strength is its flexibility (and community :P) but I think it is worth knowing where we are paying a price for flexibility, how much it is and where we are plainly inefficient and why.

**Reply by RiccardoRossi** [reply]
Per regards the linear solver, I see no reason for which we should be slower than any other solver (commercial or not).

Regarding the element assembly we are accepting some tradoffs. I indicate here the ones i consider more onerous
1 - we need dynamic matrices to communicate with elements (although the elements can use internally  stack allocated matrices). 
2 - The Constitutive Law interface which as of now implies using dynamic matrices too.  We have a proposal for correcting this, however it will imply changing quite a lot of stuff.
3 - our nodal database is flexible and this is intrinsically slower than a fixed one where at compile time it is known where the things are
4 - dofs retrieval is pretty slow since as of now it is completely dynamic.
5 - our bounded_matrix is stack allocated but has a variable_size. This has advantages and disadvantags, however (i,j) operator has some overhead with respect to optimal stack-based allocation
6 - Ublas is VERY slow compared to alternatives. @pooyan-dadvand you benchmarked this, can u comment?  
7 - If we know exactly what to do there are some ways to cut corners. For example one may use lower order integration for some terms than for others. For example for a Tet element one may even avoid integration loops alltogether.

having said this, i am pretty sure that one could obtain a 2-3 fold speedup FOR THE ASSEMBLY of some specific elements (of course not for every element, depending on the starting quaility of the implementation) without any change in the structure of Kratos and by simply taking some more care (surely at the price of some internal flexibility) of how the elements are written. 

Other than that points 2,4,5,6 could be attacked programmatically (not to say 7 of course). This however would imply some workforce/funding (jeje, pinging the industry guys here) and some tolerance to disruption from the user community

**Reply by jrubiogonzalez** [reply]
Interesting, if x2-3 speed up in assembly can be achieved probably would be good to document the best practices. I'm pretty sure that we are overlooking other bottlenecks. Ideas on how to profile and spot them? Just to see what is achievable.

**Reply by loumalouomega** [reply]
> 6 - Ublas is VERY slow compared to alternatives. @pooyan-dadvand you benchmarked this, can u comment?

Blaze benchmarks:

https://bitbucket.org/blaze-lib/blaze/wiki/Benchmarks

(SPOILER: Blaze wins everything)

**Reply by philbucher** [reply]
> > 6 - Ublas is VERY slow compared to alternatives. @pooyan-dadvand you benchmarked this, can u comment?
> 
> Blaze benchmarks:
> 
> https://bitbucket.org/blaze-lib/blaze/wiki/Benchmarks
> 
> (SPOILER: Blaze wins everything)

there is also a [talk from the author of Blaze from the cppcon 2016](https://www.youtube.com/watch?v=w-Y22KrMgFE). I am a fan of him, had some courses with him, he seems to really know what he is doing

**Reply by pooyan-dadvand** [reply]
Thanks to @RiccardoRossi for generating an extensive list of Kratos core bottlenecks coming from the flexibility!! And indeed, they are not few!

Of course, finding the sweet spot in a trad-off situation like this is not very easy. I think we use to do it well and in many areas (like structural mechanics) we find a good compromise. But, every day we are opening to new solutions (like explicit ones) in which our balance is not valid anymore. 

I think maybe it is time to add a new way of adding physics to the Kratos (besides the current elements) which is more appropriate for the solutions that are less flexible by nature but more on the performance side. (Here I am thinking on DEM, explicit, and some fluids) For example to have the possibility of working with several same geometries together without the overhead of the dynamic allocations mentioned in many parts. Of course, here the problem is to make it compatible with the rest of the Kratos but I think it is doable. 

Now the question is if we have enough interest and resources to do it?

**Reply by mpentek** [reply]
Apart from the very technical aspects listed, how does one properly compare for example Kratos FluidDyn and OpenFOAM CFD? As a user one is interested in the effort (human+machine) for a certain degree of quality of the results.

Now omitting the human part, how do we compare it fairly?
The whole workflow (from generating usable meshes to solving)?
Comparing the number of DoFs and global matrix sizes (FEM vs FVM discussion?)?
Comparing the spatial and temporal resolution (time step sensitivity of stability and accuracy)?
...

At least for me, even coming up with a fair (and usable) comparing scenario is hard. This only means in my view that one can and probably should focus on improvements in Kratos (which can be profiled/measured), whereas thinking about an objective (and useful) comparison with other software is very tricky. Or one can always plan the comparison such that the "winner" is known already when the "race" starts.

**Comment by RiccardoRossi** [comment]
well, vtune is your friend...

**Reply by loumalouomega** [reply]
And now is free 👯

**Reply by jrubiogonzalez** [reply]
vtune and friend in the same sentence...
![image](https://user-images.githubusercontent.com/25137726/110517376-12810200-810b-11eb-94b6-6389441553f1.png)

vtune is the tool, but how to profile each of your proposals to know the impact is the hard part.

**Comment by vryy** [comment]
I think the (most) advantage of Kratos is its flexibility to combine
different components to make an analysis and we should really praise it. In
terms of efficiency, to profile and optimize a code can take more time than
developing a physics feature, and sometimes your solution is not the slow
path of the code but the correct solution strategy to converge. If you
design an analysis solution for a million engineer to use then it makes
sense to profile and optimize, writing code for a publication ... probably
not a good(best) idea.

On top of that, an idea for huge improvement in speed is to make a kind of
Compiler similar to Fenics (Dolfin). Based on the python code, the input,
confs, etc, generating an optimized executable using template-based
re-generation, vectorization, cache-aware, etc  . That will be very useful
for those deploying the solution for industry problems where you need
speed. I hope this is not a stupid one :))

**Comment by RiccardoRossi** [comment]
btw, i ordered today an epyc server with 24 cores to compile the wheels

**Comment by antonialarese** [comment]
Dear @KratosMultiphysics/all

The recorded video of the 2nd KratosCommunityMeeting as well as the respective slides are available online.

Topics addressed:

- How the CI (continuous integration) works 
- The future of shared memory parallelization in Kratos: The transition from OpenMP to C++ Threads 
- The future of C++ testing in Kratos

Video of the meeting [Link to video](https://mediaspace.unipd.it/media/2nd+KRATOS+Community+Meeting/1_27ebjrnb)
Pdf of the presentation by @philbucher [Link to slides](https://drive.google.com/file/d/1Ktbcja214ZXrHZqjqZuTAm1yLr7DpJew/view?usp=sharing)
